sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/AmqpReceiveLinkProcessor.java

```diff
@@ -41,6 +41,7 @@
     private final AtomicBoolean isTerminated = new AtomicBoolean();
     private final AtomicInteger retryAttempts = new AtomicInteger();
     private final Deque<Message> messageQueue = new ConcurrentLinkedDeque<>();
+    private final AtomicBoolean linkHasNoCredits = new AtomicBoolean();
```

Comment:
jfyi, Ideally a scalar boolean should do if access to `linkHasNoCredits` is always guarded by `creditsAdded`; the aquire-release (a-r) semantic provided by `creditsAdded` should ensure modification of all shared data sandwiched between the a-r is visible to threads in all cpus.

Reply:
We might be able to go further to remove this attribute

Reply:
There are times we asynchronously update it outside of a `creditsAdded` lock. For example, our usage in `next.setEmptyCreditListener` and some of the subscriber callbacks when an exception occurs asynchronously.